### ISSUE TITLE
Scoping a `has_one` relationship

### ISSUE THREAD

**Issue by davidgoli** [JSONAPI-Resources/jsonapi-resources#627]
After searching for a while, the answer here isn't obvious to me, so I thought I'd bring it up as an issue for reference.

I have 3 models:

``` ruby
class User < ActiveRecord::Base
  has_many :answers
end

class Question < ActiveRecord::Base
  has_many :answers
end

class Answer < ActiveRecord::Base
  belongs_to :user
  belongs_to :question
end
```

What I'd like to do is `include` the `current_user_answer` in a `questions#show` request, like so: `/api/questions/123?include=current_user_answer`. It seems like I should be able to write my `QuestionResource` as follows:

``` ruby
class QuestionResource < JSONAPI::Resource
  has_many :answers

  has_one :current_user_answer, class_name: 'Answer'

  # doesn't get called
  def self.record_for_current_user_answer(options={})
        current_user = options[:context][:current_user]
        AnswerResource.records(options).find_by(
          question: @model,
          user: current_user
        )
  end

  # doesn't get called either
  def current_user_answer(options={})
        current_user = options[:context][:current_user]
        AnswerResource.records(options).find_by(
          question: @model,
          user: current_user
        )
  end
end
```

Neither of these methods get called; instead, I get an exception `"Association named 'current_user_answer' was not found on Question; perhaps you misspelled it?"` Well, by definition here I'm trying to avoid writing a query method directly on the model, and anyways such a query method or association would need to get the `context` passed in from the enclosing Resource.

In short, there seems to be no documented way to scope an `include`d associated to something in the `context`. Or am I missing something? What's the canonical way to get this related resource without an extra request?

**Comment by davidgoli** [comment]
I think this is a similar issue to https://github.com/cerebris/jsonapi-resources/issues/545. In my case, the associated record IS actually backed by a record, it's just that the query needs additional filtering to find it.

**Comment by lgebhardt** [comment]
@davidgoli Are you sure `current_user_answer` doesn't get called? I just modeled this and it's being called in my code. However the find code should be:

``` ruby
def current_user_answer(options={})
  current_user = options[:context][:current_user]
  AnswerResource.records(options).where(question: @model.id, user: current_user)
end
```

This will return an ActiveRecord Relation instead of an instance of the model.

**Comment by davidgoli** [comment]
Yes, I'm sure. I can set a `binding.pry` in this `current_user_answer` method and it never gets hit. Instead, I continue to get `Internal Server Error: Association named 'current_user_answer' was not found on Question; perhaps you misspelled it? /Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations.rb:154:in`association'`

Perhaps one wrinkle here is I'm overriding `self.find_by_key` because I have a special key value `of-the-day` which basically means "latest", so my full stack trace is:

```
Internal Server Error: Association named 'current_user_answer' was not found on Question; perhaps you misspelled it? /Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations.rb:154:in `association'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:147:in `block in grouped_records'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:145:in `each'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:145:in `grouped_records'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:134:in `preloaders_for_one'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:106:in `preloaders_on'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:94:in `block in preload'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:93:in `each'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:93:in `flat_map'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/associations/preloader.rb:93:in `preload'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation.rb:617:in `block in exec_queries'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation.rb:616:in `each'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation.rb:616:in `exec_queries'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation.rb:493:in `load'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation.rb:238:in `to_a'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/finder_methods.rb:476:in `find_nth_with_limit'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/finder_methods.rb:468:in `find_nth'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/finder_methods.rb:132:in `first'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/finder_methods.rb:139:in `first!'
/Users/davidgolightly/dev/Siren/app/models/question.rb:14:in `of_the_day'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/delegation.rb:94:in `public_send'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/delegation.rb:94:in `block in method_missing'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation.rb:292:in `scoping'
/Users/davidgolightly/.rvm/gems/ruby-2.2.3@siren/gems/activerecord-4.1.13/lib/active_record/relation/delegation.rb:94:in `method_missing'
/Users/davidgolightly/dev/Siren/app/resources/api/v3/question_resource.rb:57:in `find_by_key'
```

**Comment by davidgoli** [comment]
My implementation of `find_by_key` is:

``` ruby
      def self.find_by_key(id, options={})
        context = options[:context]
        records = self.records(options)
        binding.pry
        records = self.apply_includes(records, options)

        model = if id == 'of-the-day'
          records.of_the_day
        else
          records.find(id)
        end

        fail JSONAPI::Exceptions::RecordNotFound.new(key) if model.nil?
        resource_for_model(model).new(model, context)
      end
```

And by the way, this is all in an attempt to convert an existing REST API to the JSONAPI spec for easier use with Ember.JS. I've found it very easy to adopt this gem on a greenfield project, it's adapting existing structures like this that are a little tricky.

**Comment by davidgoli** [comment]
I looks like what's happening here is in `jsonapi/resource.rb:490`:

``` ruby
      def apply_includes(records, options = {})
        include_directives = options[:include_directives]
        if include_directives
          model_includes = resolve_relationship_names_to_relations(self, include_directives.model_includes, options)
          records = records.includes(model_includes)
        end

        records
      end
```

in short, the `include` directives are just getting slapped on via a naive `includes` directive, which as far as I can tell doesn't hit the `resource` code. I think I'll try getting around this by singling out the `current_user_answer` key in the `includes` then special-casing it, but I wish it didn't have to be that hairy...

**Comment by davidgoli** [comment]
though, I don't think my implementation of `find_by_key` is necessarily to blame - when I change it to this, nothing changes:

``` ruby
      def self.find_by_key(id, options={})
        if id == 'of-the-day'
          id = Question.of_the_day.id
        end

        super(id, options)
      end
```
